File: src/main/java/com/gdep/Commands.java
package com.gdep;

import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class Commands {
    public static class Dirs implements Command {
        @Override
        public String getName() {
            return "dirs";
        }

        @Override
        public String getDescription() {
            return "list source directories";
        }

        @Override
        public void run(List<SourceCode> sourceCodes, String[] args) {
            List<String> sourceDirs =
                    sourceCodes.stream().map(x -> x.sourceDirPath()).sorted().toList();

            for (final String dir : sourceDirs) {
                System.out.println(dir);
            }
        }
    }

    public static class Files implements Command {
        @Override
        public String getName() {
            return "files";
        }

        @Override
        public String getDescription() {
            return "list source files";
        }

        @Override
        public void run(List<SourceCode> sourceCodes, String[] args) throws IOException {
            List<String> files = new ArrayList<>();

            for (final SourceCode code : sourceCodes) {
                files.addAll(code.sourceFiles());
            }

            files.sort(Comparator.naturalOrder());

            for (String file : files) {
                System.out.println(file);
            }
        }
    }

    public static class Pack implements Command {
        @Override
        public String getName() {
            return "pack";
        }

        @Override
        public String getDescription() {
            return "search files using class path. e.g. gdep pack o.s.w.s.DispatcherServlet";
        }

        @Override
        public void run(List<SourceCode> sourceCodes, String[] args) throws IOException {
            if (args.length <= 0) {
                throw new GracefulException("pack command needs atleast one argument").withHelp();
            }

            record PathAndScore(String path, int score) {}

            List<PathAndScore> pathAndScores = new ArrayList<>();

            for (final SourceCode sc : sourceCodes) {
                for (final String file : sc.sourceFiles()) {
                    String scoreFileName = file;

                    if (scoreFileName.startsWith(sc.sourceDirPath())) {
                        scoreFileName =
                                scoreFileName.substring(sc.sourceDirPath().length(), scoreFileName.length());
                        Path scoreFilePath = Path.of(scoreFileName);
                        if (scoreFilePath.getNameCount() >= 2) {
                            scoreFileName = scoreFilePath
                                    .subpath(1, scoreFilePath.getNameCount())
                                    .toString();
                        }
                    }

                    int score = PackScore.scoreClassNameSimilarity(scoreFileName, args[0]);
                    pathAndScores.add(new PathAndScore(file, score));
                }
            }

            List<PathAndScore> candidates = pathAndScores.stream()
                    .sorted(Comparator.comparingInt(PathAndScore::score)
                            .reversed()
                            .thenComparing(PathAndScore::path))
                    .limit(5)
                    .toList();

            for (final PathAndScore candidate : candidates) {
                System.out.println(candidate.path());
            }
        }
    }
}

File: go_wrapper/commands.go
package main

import (
	"fmt"
	"path/filepath"
	"slices"
	"strings"
	"sync"
	"flag"
	"os"
)

type Command interface {
	GetName() string
	GetDescription() string

	ParseArgs(args []string) error

	Run(sourceCodes []SourceCode) error
}

// ==============================
// Dir
// ==============================

type DirsCommand struct {
}

func (d *DirsCommand) GetName() string {
	return "dirs"
}

func (d *DirsCommand) GetDescription() string {
	return "list source directories"
}

func (d *DirsCommand) ParseArgs(args []string) error {
	return GetDefaultFlagSet(d).Parse(args)
}

func (d *DirsCommand) Run(sourceCodes []SourceCode) error {
	slices.SortFunc(sourceCodes, func(a, b SourceCode) int {
		return strings.Compare(a.GetSourceDirPath(), b.GetSourceDirPath())
	})

	for _, sc := range sourceCodes {
		fmt.Println(sc.GetSourceDirPath())
	}

	return nil
}

// ==============================
// Files
// ==============================

type FilesCommand struct {
}

func (f *FilesCommand) GetName() string {
	return "files"
}

func (f *FilesCommand) GetDescription() string {
	return "list source files"
}

func (f *FilesCommand) ParseArgs(args []string) error {
	return GetDefaultFlagSet(f).Parse(args)
}

func (f *FilesCommand) Run(sourceCodes []SourceCode) error {
	srcFiles := make([]string, 0, 1024)

	for _, sc := range sourceCodes {
		srcFiles = append(srcFiles, sc.GetSourceFiles()...)
	}

	slices.Sort(srcFiles)

	for _, file := range srcFiles {
		fmt.Println(file)
	}

	return nil
}

// ==============================
// Pack
// ==============================

type PackCommand struct {
	ClassName string
	CandidateCount uint
}

func (p *PackCommand) GetName() string {
	return "pack"
}

func (p *PackCommand) GetDescription() string {
	return "search files using class path. e.g. gdep pack o.s.w.s.DispatcherServlet"
}

func (p *PackCommand) ParseArgs(args []string) error {
	flagset := GetDefaultFlagSet(p)

	flagset.UintVar(&p.CandidateCount, "n", 5, "number of candidates to print")

	if err:= flagset.Parse(args); err != nil {
		return err
	}

	args = flagset.Args()

	if len(args) <= 0 {
		fmt.Fprintf(os.Stderr, "pack command needs atleast one argument\n")
		flagset.Usage()
		return ErrExpected
	}
	p.ClassName = strings.TrimSpace(args[0])

	return nil
}

func (p *PackCommand) Run(sourceCodes []SourceCode) error {
	type pathAndScore struct {
		Dir   string
		Path  string
		Score int
	}

	pathAndScores := make([]pathAndScore, 0, 1024)

	for _, sc := range sourceCodes {
		for _, file := range sc.GetSourceFiles() {
			pathAndScores = append(pathAndScores, pathAndScore{
				Dir:   sc.GetSourceDirPath(),
				Path:  file,
				Score: 0,
			})
		}
	}

	cursor := 0
	batchSize := 100

	var wg sync.WaitGroup

	for cursor < len(pathAndScores) {
		begin := cursor
		end := min(cursor+batchSize, len(pathAndScores))
		wg.Add(1)

		go func() {
			defer wg.Done()

			mat := NewMatrix(512)

			for i := begin; i < end; i++ {
				scoreFileName := pathAndScores[i].Path
				if rel, relErr := filepath.Rel(pathAndScores[i].Dir, scoreFileName); relErr == nil {
					scoreFileName = rel
				}

				score := ScoreClassNameSimilarity(scoreFileName, p.ClassName, mat)
				pathAndScores[i].Score = score
			}
		}()

		cursor += batchSize
	}

	wg.Wait()

	slices.SortFunc(pathAndScores, func(a, b pathAndScore) int {
		if a.Score > b.Score {
			return -1
		} else if a.Score < b.Score {
			return 1
		} else {
			return 0
		}
	})

	for i := 0; i < min(len(pathAndScores), int(p.CandidateCount)); i++ {
		ps := pathAndScores[i]
		fmt.Println(ps.Path)
	}

	return nil
}

// ==============================
// Helper Functions
// ==============================

func GetDefaultFlagSet(cmd Command) *flag.FlagSet{
	flagset := flag.NewFlagSet(cmd.GetName(), flag.ExitOnError)

	flagset.Usage = func() {
		fmt.Fprintf(os.Stderr, "Command %s:\n", cmd.GetName())
		fmt.Fprintf(os.Stderr, "  %s\n", cmd.GetDescription())
		flagset.PrintDefaults()
	}

	return flagset;
}
